### PR TITLE
CNS-923: added local cache for validate pairing

### PR DESCRIPTION
**Pull request by oren-lava**
When iterating over relays, instead of validating the pairing for each relay, we keep track of the validated provider-consumer-spec trio in a local map and skip the validation if it was already validated in a previous relay

### CODE REVIEW

File: x/pairing/keeper/msg_server_relay_payment.go [hunk 1]
```diff
@@ -169,24 +176,36 @@ func (k msgServer) RelayPayment(goCtx context.Context, msg *types.MsgRelayPaymen
...
+		var providers []epochstoragetypes.StakeEntry
+		allowedCU := uint64(0)
+		pairingData, ok := validatePairingCache[validatePairingKey]
+		if ok {
+			allowedCU = pairingData.allowedCU
+			providers = pairingData.providers
+		} else {
+			isValidPairing := false
+			isValidPairing, allowedCU, providers, err = k.Keeper.ValidatePairingForClient(
+				ctx,
```
Comment (Yaroms) [inline x/pairing/keeper/msg_server_relay_payment.go:190]: can allowed CU change in between iterations? if so we need to recalculate it

Comment (oren-lava) [reply x/pairing/keeper/msg_server_relay_payment.go:193]: done. see https://github.com/lavanet/lava/commit/e33ae32dd77042944191f696e87d51636ed548db

File: x/pairing/keeper/msg_server_relay_payment.go [hunk 2]
```diff
@@ -271,6 +294,9 @@ func (k msgServer) RelayPayment(goCtx context.Context, msg *types.MsgRelayPaymen
 		rejected_relays_num--
 	}
 
+	// save all payment objects
+	paymentHandler.Flush(ctx)
```
Comment (oren-lava) [inline x/pairing/keeper/msg_server_relay_payment.go:298]: wouldn't it be better if we put the flush after the `rejected_relays_num != 0` check? In other words, we don't wanna update the KVStore for a failed relay TX

File: x/pairing/keeper/unique_payment_storage_client_provider.go [hunk 3]
```diff
@@ -73,8 +73,8 @@ func (k Keeper) IsDoubleSpend(ctx sdk.Context, chainID string, block uint64, pro
 	return found
 }
 
-func (k Keeper) GetConsumerFromUniquePayment(uniquePaymentStorageClientProvider *types.UniquePaymentStorageClientProvider) string {
-	key := uniquePaymentStorageClientProvider.Index
+func (k Keeper) GetConsumerFromUniquePayment(uniquePaymentStorageClientProvider string) string {
+	key := uniquePaymentStorageClientProvider
```
Comment (oren-lava) [inline x/pairing/keeper/unique_payment_storage_client_provider.go:77]: name the function's argument `key` and remove this

**Review by Yaroms — APPROVED** [review]

**Pull request merged**